File: src/UnMango.Rest/RestRequest.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using UnMango.Rest.Internal;

namespace UnMango.Rest
{
    /// <summary>
    /// Represents a REST request.
    /// </summary>
    public sealed class RestRequest : IRestRequest
    {
        /// <summary>
        /// Initializes a new instance of a <see cref="RestRequest"/> that will use
        /// <paramref name="client"/> to execute a REST request.
        /// </summary>
        /// <param name="client">The <see cref="IRestClient"/> to use to execute the request.</param>
        public RestRequest(IRestClient client)
        {
            Client = client ?? throw new ArgumentNullException(nameof(client));
            QueryParameters = new Dictionary<string, string>();
        }

        /// <summary>
        /// Initializes a new instance of a <see cref="RestRequest"/> that will use
        /// <paramref name="client"/> to execute a <paramref name="method"/> request.
        /// </summary>
        /// <param name="client">The <see cref="IRestClient"/> to use to execute the request.</param>
        /// <param name="method">The <see cref="HttpMethod"/> the request will use.</param>
        public RestRequest(IRestClient client, HttpMethod method) : this(client)
        {
            Method = method;
            QueryParameters = new Dictionary<string, string>();
        }

        /// <summary>
        /// Initializes a new instance of a <see cref="RestRequest"/> that will use
        /// <paramref name="client"/> to execute a <paramref name="method"/> request
        /// with <paramref name="queryParameters"/> as parameters.
        /// </summary>
        /// <param name="client">The <see cref="IRestClient"/> to use to execute the request.</param>
        /// <param name="method">The <see cref="HttpMethod"/> the request will use.</param>
        /// <param name="queryParameters">The query parameters to send with the request.</param>
        public RestRequest(
            IRestClient client,
            HttpMethod method,
            IEnumerable<KeyValuePair<string, string>> queryParameters)
            : this(client, method)
        {
            if (queryParameters == null) throw new ArgumentNullException(nameof(queryParameters));
            
            QueryParameters = new Dictionary<string, string>(queryParameters.ToDictionary());
        }
        
        /// <inheritdoc />
        public IRestClient Client { get; }

        /// <inheritdoc />
        public HttpMethod? Method { get; }

        /// <inheritdoc />
        public IReadOnlyDictionary<string, string> QueryParameters { get; }

        /// <inheritdoc />
        public string Uri { get; }
    }
}

File: src/UnMango.Rest/RestRequest.fs
namespace UnMango.Rest

open System
open System.Net.Http
open System.Threading

type RestRequest<'B, 'R> =
    { CancellationToken: CancellationToken option
      Body: 'B option
      Client: IRestClient
      Method: HttpMethod option
      QueryParameters: Map<string, string>
      Uri: Uri option }

    member this.GetAwaiter() =
        match this.CancellationToken with
        | Some token -> this.Client.SendAsync(this, token).GetAwaiter()
        | None -> this.Client.SendAsync(this).GetAwaiter()

    interface IRestRequest with
        member this.Client = this.Client
        member this.Method = this.Method |> Option.get
        member this.QueryParameters = this.QueryParameters
        member this.GetAwaiter() = this.GetAwaiter()

module RestRequest =
    let internal unwrap (r: IRestRequest) =
        match r with
        | :? RestRequest<_, _> as req -> req
        | _ ->
            { CancellationToken = None
              Body = None
              Client = r.Client
              Method = Some r.Method
              QueryParameters = r.QueryParameters |> Util.toMap
              Uri = None }

    let Create(c: HttpClient) : IRestRequest =
        { CancellationToken = None
          Body = None
          Client = RestClient(c)
          Method = None
          QueryParameters = Map.empty
          Uri = None }
